Source/SaveStates/SceneWatcher.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using UnityEngine.SceneManagement;
+using Object = UnityEngine.Object;
+using USceneManager = UnityEngine.SceneManagement.SceneManager;
+
+namespace DebugMod
+{
+    public static class SceneWatcher
+    {
+        private static List<SceneData> scenes;
+        public static ReadOnlyCollection<SceneData> LoadedScenes => scenes.AsReadOnly();
+
+        private static Dictionary<Scene, int> scenesWithManager;
+
+        private static void AddScene(Scene scene, LoadSceneMode mode, bool checkSceneManager = true)
+        {
+            if (mode == LoadSceneMode.Single)
+                scenes.Clear();
+
+            SceneData d = new SceneData(scene.name, scene.name);
+            scenes.Add(d);
+
+            if (checkSceneManager && Object.FindObjectsOfType<SceneManager>().Any(m => m.gameObject.scene == scene))
+                scenesWithManager.Add(scene, d.id);
+        }
+
+        public static void Init()
+        {
+            scenesWithManager = new();
+            scenes = new();
+
+            for (int i = 0; i < USceneManager.sceneCount; i++)
+                AddScene(USceneManager.GetSceneAt(i), LoadSceneMode.Additive, false);
+            
+            USceneManager.sceneLoaded += (scene, mode) => AddScene(scene, mode);
+            On.SceneManager.Start += (orig, self) =>
+            {
+                orig(self);
+
+                if (!scenesWithManager.ContainsKey(self.gameObject.scene))
+                    return;
+
+                int id = scenesWithManager[self.gameObject.scene];
+                SceneData data = scenes.FirstOrDefault(d => d.id == id);
+                scenesWithManager.Remove(self.gameObject.scene);
+                
+                if (data != null)
+                    data.activeSceneWhenLoaded = GameManager.instance.sceneName;
+            };
+            USceneManager.sceneUnloaded += s => scenes.RemoveAt(scenes.FindIndex(d => d.name == s.name));
+        }
+        
+        public class SceneData
```

Comment:
Perhaps this class should not be called SceneData, to confuse with the SceneData class in HK

Reply:
Loaded Scene Data?

Reply:
I think I like info more than data

Reply:
Sure, can commit tomorrow